File: backend/app/schemas/roadmap.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from pydantic import BaseModel, ConfigDict, Field

RoadmapMode = Literal["strict", "adaptive", "hybrid"]
NodeStatus = Literal["not_started", "in_progress", "completed", "skipped", "flagged"]
NodeType = Literal["subject", "chapter", "topic", "bridge"]
EdgeType = Literal["sequential", "prerequisite", "bridge"]


class RoadmapCreate(BaseModel):
    document_id: str | None = Field(None, description="ID of syllabus document to build roadmap from")
    mode: RoadmapMode = Field("strict", description="Roadmap generation mode: strict, adaptive, or hybrid")
    title: str | None = Field(None, description="Optional custom title for the roadmap")
    syllabus_text: str | None = Field(None, description="Optional raw syllabus text instead of document ID")


class RoadmapNodeUpdate(BaseModel):
    status: NodeStatus | None = None
    mastery_score: float | None = Field(None, ge=0.0, le=1.0)
    time_spent_minutes: int | None = Field(None, ge=0)


class RoadmapNodeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    roadmap_id: str
    profile_id: str
    title: str
    description: str | None = None
    node_type: str
    status: str
    parent_id: str | None = None
    order_index: int
    mastery_score: float
    time_spent_minutes: int
    ai_generated: bool
    completed_at: datetime | None = None
    unlocked: bool = True


class RoadmapEdgeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    roadmap_id: str
    from_node_id: str
    to_node_id: str
    edge_type: str


class RoadmapProgress(BaseModel):
    total_nodes: int = 0
    completed_nodes: int = 0
    in_progress_nodes: int = 0
    completion_percentage: float = 0.0
    average_mastery: float = 0.0


class RoadmapSummaryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    profile_id: str
    title: str
    mode: str
    version: int
    is_active: bool
    source_document_id: str | None = None
    created_at: datetime
    node_count: int = 0


class RoadmapResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    profile_id: str
    title: str
    mode: str
    version: int
    is_active: bool
    source_document_id: str | None = None
    created_at: datetime
    nodes: list[RoadmapNodeResponse] = Field(default_factory=list)
    edges: list[RoadmapEdgeResponse] = Field(default_factory=list)
    progress: RoadmapProgress | None = None


# Alias
RoadmapDetailResponse = RoadmapResponse

File: frontend/src/api/roadmap.ts
import { apiClient, ApiEnvelope } from "./client";

export interface RoadmapNode {
  id: string;
  roadmap_id: string;
  profile_id: string;
  title: string;
  description?: string | null;
  node_type: "subject" | "chapter" | "topic" | "bridge";
  status: "not_started" | "in_progress" | "completed" | "skipped" | "flagged";
  parent_id?: string | null;
  order_index: number;
  mastery_score: number;
  time_spent_minutes: number;
  ai_generated: boolean;
  completed_at?: string | null;
  unlocked: boolean;
}

export interface RoadmapEdge {
  id: string;
  roadmap_id: string;
  from_node_id: string;
  to_node_id: string;
  edge_type: "sequential" | "prerequisite" | "bridge";
}

export interface RoadmapProgress {
  total_nodes: number;
  completed_nodes: number;
  in_progress_nodes: number;
  completion_percentage: number;
  average_mastery: number;
}

export interface RoadmapDetail {
  id: string;
  profile_id: string;
  title: string;
  mode: "strict" | "adaptive" | "hybrid";
  version: number;
  is_active: boolean;
  source_document_id?: string | null;
  created_at: string;
  nodes: RoadmapNode[];
  edges: RoadmapEdge[];
  progress?: RoadmapProgress | null;
}

export interface RoadmapSummary {
  id: string;
  profile_id: string;
  title: string;
  mode: string;
  version: number;
  is_active: boolean;
  source_document_id?: string | null;
  created_at: string;
  node_count: number;
}

export const roadmapApi = {
  listRoadmaps: async (profileId: string): Promise<RoadmapSummary[]> => {
    const res = await apiClient.get<ApiEnvelope<RoadmapSummary[]>>(
      `/profiles/${profileId}/roadmaps`
    );
    return res.data.data;
  },

  getActiveRoadmap: async (profileId: string): Promise<RoadmapDetail> => {
    const res = await apiClient.get<ApiEnvelope<RoadmapDetail>>(
      `/profiles/${profileId}/roadmaps/active`
    );
    return res.data.data;
  },

  getRoadmap: async (profileId: string, roadmapId: string): Promise<RoadmapDetail> => {
    const res = await apiClient.get<ApiEnvelope<RoadmapDetail>>(
      `/profiles/${profileId}/roadmaps/${roadmapId}`
    );
    return res.data.data;
  },

  generateRoadmap: async (
    profileId: string,
    payload: {
      document_id?: string;
      /** Course materials to write a curriculum from, when no syllabus exists. */
      document_ids?: string[];
      mode?: "strict" | "adaptive" | "hybrid";
      title?: string;
      syllabus_text?: string;
    }
  ): Promise<RoadmapDetail> => {
    const res = await apiClient.post<ApiEnvelope<RoadmapDetail>>(
      `/profiles/${profileId}/roadmaps`,
      payload
    );
    return res.data.data;
  },

  updateNodeStatus: async (
    profileId: string,
    roadmapId: string,
    nodeId: string,
    payload: {
      status?: "not_started" | "in_progress" | "completed" | "skipped" | "flagged";
      mastery_score?: number;
      time_spent_minutes?: number;
    }
  ): Promise<RoadmapNode> => {
    const res = await apiClient.patch<ApiEnvelope<RoadmapNode>>(
      `/profiles/${profileId}/roadmaps/${roadmapId}/nodes/${nodeId}`,
      payload
    );
    return res.data.data;
  },

  regenerateRoadmap: async (
    profileId: string,
    roadmapId: string
  ): Promise<RoadmapDetail> => {
    const res = await apiClient.post<ApiEnvelope<RoadmapDetail>>(
      `/profiles/${profileId}/roadmaps/${roadmapId}/regenerate`
    );
    return res.data.data;
  },
};
